### PR TITLE
Refactor `repr` and `str` of `QuadraticProgram` and `OptimizationResult` (part 3/3 refactor text representation of objects)

### LINKED ISSUE #77: Add textual representations to optimization objects
### What is the expected enhancement?
Many optimization objects, e.g. `Constraint`, `Variable` do not have textual representations, e.g. methods like `__str__` and `__repr__` and in a debugger, e.g. PyCharm they don't provide much information out of box. `QuadraticProgram` has only `__repr__` which, in turns, depends on the docplex implementation, and in the debugger you can only see a docplex comment header of the LP representation.

This issue should be fixed by reviewing optimization objects and introducing textual representation methods in them.

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

- Adds `qiskit_optimization.algorithms.OptimizationResult.prettyprint` to display a multi-line text.
- Adds `wrap` option to `prettyprint` function and method to set the text width.
- Updates `str` and `repr` of `QuadraticProgram` and `OptimizationResult` to be a one-line text.
- Reword "optimal" in `OptimizationResult` and `OptimizationResultStatus` because some algorithms do not guarantee the optimality.
  - "optimal solution" (`OptimizationResultStatus.SUCCESS`) -> "feasible solution"
  - "optimal function value" (`OptimizationResult.fval`) -> "objective function value"
  - "optimal value" (`OptimizationResult.x`) -> "variable values"

TODO
- [x] Refactors `repr` and `str` of `QuadraticProgram` and `OptimizationResult` to be one line
- [x] Adds `OptimizationResult.prettyprint` to generate a multi-line string
- [x] Updates docstrings of `OptimizationResults` and `OptimizationResultStatus`
- [x] Updates tutorials to replace `export_as_lp_string` with `prettyprint`
- [x] Updates reno 

Closes #184
Closes #77

Follow-up of #366 #369

### Details and comments

`QuadraticProgram`
```python
from qiskit_optimization.problems import QuadraticProgram

qp = QuadraticProgram('problem 1')
qp.integer_var(-1, 2, 'x')
qp.integer_var(-1, 2, 'y')
qp.continuous_var(-1, name='z')
qp.minimize(constant=3, linear={'x': 2, 'y': 3}, quadratic={('z', 'x'): -1})
qp.linear_constraint({'x': 1, 'y': -2}, '>=', 2, name='lin_GE')
qp.linear_constraint({'x': 2, 'y': -1}, '==', 1, name='lin_EQ')
qp.quadratic_constraint({'x': 2, 'y': -1}, {('z', 'z'): -1}, '<=', 1)

print(repr(qp))
print()
print(str(qp))
print()
print(qp.prettyprint())
```

```
<QuadraticProgram: minimize -x*z + 2*x + 3*y + 3, 3 variables, 3 constraints, 'problem 1'>

minimize -x*z + 2*x + 3*y + 3 (3 variables, 3 constraints, 'problem 1')

Problem name: problem 1

Minimize
  -x*z + 2*x + 3*y + 3

Subject to
  Linear constraints (2)
    x - 2*y >= 2  'lin_GE'
    2*x - y == 1  'lin_EQ'

  Quadratic constraints (1)
    -z^2 + 2*x - y <= 1  'q0'

  Integer variables (2)
    -1 <= x <= 2
    -1 <= y <= 2

  Continuous variables (1)
    -1 <= z
```
-----------------------

`OptimizationResult`

```python
from qiskit_optimization.problems import QuadraticProgram
from qiskit_optimization.algorithms import OptimizationResult, OptimizationResultStatus

qp = QuadraticProgram()
x = qp.binary_var_list(3)
result = OptimizationResult([1.0,2.0,3.0], 10.0, x, OptimizationResultStatus.SUCCESS)

print(str(result))
print()
print(repr(result))
print()
print(str(result.prettyprint()))
```

```
fval=10.0, x0=1.0, x1=2.0, x2=3.0, status=SUCCESS

<OptimizationResult: fval=10.0, x0=1.0, x1=2.0, x2=3.0, status=SUCCESS>

objective function value: 10.0
variable values: x0=1.0, x1=2.0, x2=3.0
status: SUCCESS
```

### CODE REVIEW

File: docs/tutorials/01_quadratic_program.ipynb [hunk 1]
```diff
@@ -63,7 +63,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "You start with an empty model. How to add variables and constraints to a model is explained in the section \"Directly constructing a `QuadraticProgram`\"."
+    "You start with an empty model. How to add variables and constraints to a model is explained in the section [Directly constructing a `QuadraticProgram`](#Directly-constructing-a-QuadraticProgram)."
```
Comment (woodsp-ibm) [inline docs/tutorials/01_quadratic_program.ipynb:66]: I was viewing this file and the link seemed not to work as expected. It certainly fails to come out in the html documentation where I'd argue its more important there than in the source 
image

File: qiskit_optimization/algorithms/optimization_algorithm.py [hunk 2]
```diff
@@ -123,7 +123,7 @@ def __init__(
         else:
             if len(x) != len(variables):
                 raise QiskitOptimizationError(
-                    f"Inconsistent size of optimal value and variables. x: size {len(x)} {x}, "
+                    f"Inconsistent size of variables and values. x: size {len(x)} {x}, "
```
Comment (woodsp-ibm) [inline qiskit_optimization/algorithms/optimization_algorithm.py:126]: In this message it looks like you change `optimal value` to `values` (`variable values` more fully above when describing x) and switched the order of that text.r with `variables`, though in the message itself it has values and variables in that order when it prints out the actual quantities that mismatch

File: docs/tutorials/02_converters_for_quadratic_programs.ipynb [hunk 3]
```diff
@@ -266,51 +242,32 @@
...
-      " 0 <= xyz_leq@int_slack@0 <= 1\n",
-      " 0 <= xyz_leq@int_slack@1 <= 1\n",
-      " 0 <= xyz_leq@int_slack@2 <= 1\n",
-      " 0 <= xyz_geq@int_slack@0 <= 1\n",
-      " 0 <= xyz_geq@int_slack@1 <= 1\n",
-      " 0 <= xyz_geq@int_slack@2 <= 1\n",
+      "Subject to\n",
+      "  Linear constraints (2)\n",
+      "    x + xyz_leq@int_slack@0 + 2*xyz_leq@int_slack@1 + 2*xyz_leq@int_slack@2 + y + z@0 + 2*z@1 + 4*z@2 == 5  'xyz_leq'\n",
+      "    x - xyz_geq@int_slack@0 - 2*xyz_geq@int_slack@1 - 3*xyz_geq@int_slack@2 + y + z@0 + 2*z@1 + 4*z@2 == 3  'xyz_geq'\n",
```
Comment (woodsp-ibm) [inline docs/tutorials/02_converters_for_quadratic_programs.ipynb:259]: These end up wrapping in the html docs so the end result is not as nice. Not sure there is a great deal one can do about this. I imagine that we could have the prettyprint wrap things nicely at a given limit, so it determines/manages the word wrapping, but maybe that's too much work for now.
image

Comment (t-imamichi) [reply docs/tutorials/02_converters_for_quadratic_programs.ipynb:259]: Thanks. Wrapping lines might be a good option to improve the appearance. I will try it.

Comment (t-imamichi) [reply docs/tutorials/02_converters_for_quadratic_programs.ipynb:259]: I'm trying `wrap` option. Here are some results with different wrap values. What do you think of these appearance? If OK, I will write docstrings  and unit tests.

wrap = 100
```
Problem name: problem 1

Minimize
   43*lin_GE@int_slack@0^2 + 86*lin_GE@int_slack@0*lin_GE@int_slack@1 + 43*lin_GE@int_slack@1^2
   - 86*x@0*lin_GE@int_slack@0 - 86*x@0*lin_GE@int_slack@1 + 215*x@0^2 + 860*x@0*x@1 - 344*x@0*y@0
   - 688*x@0*y@1 - x@0*z@0 - 2*x@0*z@1 - 3*x@0*z@2 - 172*x@1*lin_GE@int_slack@0 - 172*x@1*lin_GE@int_slack@1
   + 860*x@1^2 - 688*x@1*y@0 - 1376*x@1*y@1 - 2*x@1*z@0 - 4*x@1*z@1 - 6*x@1*z@2 + 172*y@0*lin_GE@int_slack@0
   + 172*y@0*lin_GE@int_slack@1 + 215*y@0^2 + 860*y@0*y@1 + 344*y@1*lin_GE@int_slack@0
   + 344*y@1*lin_GE@int_slack@1 + 860*y@1^2 + 86*lin_GE@int_slack@0 + 86*lin_GE@int_slack@1 - 427*x@0
   - 854*x@1 + 347*y@0 + 694*y@1 + z@0 + 2*z@1 + 3*z@2 + 212 

Subject to
  No constraints

  Binary variables (12)
     x@0 x@1 y@0 y@1 z@0 z@1 z@2 xxx3 xxx4 xxx5 lin_GE@int_slack@0 lin_GE@int_slack@1
```

wrap = 50
```
Problem name: problem 1

Minimize
   43*lin_GE@int_slack@0^2
   + 86*lin_GE@int_slack@0*lin_GE@int_slack@1
   + 43*lin_GE@int_slack@1^2
   - 86*x@0*lin_GE@int_slack@0
   - 86*x@0*lin_GE@int_slack@1 + 215*x@0^2
   + 860*x@0*x@1 - 344*x@0*y@0 - 688*x@0*y@1 - x@0*z@0
   - 2*x@0*z@1 - 3*x@0*z@2 - 172*x@1*lin_GE@int_slack@0
   - 172*x@1*lin_GE@int_slack@1 + 860*x@1^2
   - 688*x@1*y@0 - 1376*x@1*y@1 - 2*x@1*z@0 - 4*x@1*z@1
   - 6*x@1*z@2 + 172*y@0*lin_GE@int_slack@0
   + 172*y@0*lin_GE@int_slack@1 + 215*y@0^2
   + 860*y@0*y@1 + 344*y@1*lin_GE@int_slack@0
   + 344*y@1*lin_GE@int_slack@1 + 860*y@1^2
   + 86*lin_GE@int_slack@0 + 86*lin_GE@int_slack@1
   - 427*x@0 - 854*x@1 + 347*y@0 + 694*y@1 + z@0 + 2*z@1
   + 3*z@2 + 212 

Subject to
  No constraints

  Binary variables (12)
     x@0 x@1 y@0 y@1 z@0 z@1 z@2 xxx3 xxx4 xxx5
     lin_GE@int_slack@0 lin_GE@int_slack@1
```

Comment (woodsp-ibm) [reply docs/tutorials/02_converters_for_quadratic_programs.ipynb:259]: Looks good to me. Given you show different values above I assume you will wrap at some default value and let the user change that if they want for some reason.

Comment (t-imamichi) [reply docs/tutorials/02_converters_for_quadratic_programs.ipynb:259]: Thanks, I will tune the default value looking at the jupyter notebook output and let users change that if they wish.

File: docs/tutorials/01_quadratic_program.ipynb (hunk 1 above)
Comment (t-imamichi) [reply docs/tutorials/01_quadratic_program.ipynb:66]: Thanks. Avoid using the code format in the link (`QuadraticProgram` -> QuadraticProgram) resolves the issue.
image

File: qiskit_optimization/algorithms/optimization_algorithm.py (hunk 2 above)
Comment (t-imamichi) [reply qiskit_optimization/algorithms/optimization_algorithm.py]: Thanks. I updated the message as follows. I hope it is comprehensive.
```python
f"Inconsistent size of variable values (x) and variables. x: size {len(x)} {x}, "
f"variables: size {len(variables)} {[v.name for v in variables]}"
```

**Comment by t-imamichi** [comment]
I added a text wrap feature to `prettyprint` and set text width 80 as the default value.
I implemented a simple approach to wrap texts at `+` or `-` only. It does not wrap text at `*` and (in)equalities.
So, the `wrap` value is not a hard constraint (i.e., some text may exceed).
But, I think it's enough comprehensive. See https://github.com/Qiskit/qiskit-optimization/pull/370#discussion_r873707634 for examples.

**Review by woodsp-ibm — APPROVED** [review]
LGTM thx!

**Comment by t-imamichi** [comment]
Thank you for your reviews of these PR series! I finally completed them.

**Pull request merged**